cnxdb/migrations/20200506144128_update_trees_slug_values.py

```diff
@@ -0,0 +1,196 @@
+VARNISH_HOSTS = os.getenv('VARNISH_HOSTS', '').split(',')
+VARNISH_PORT = int(os.getenv('VARNISH_PORT', 80))
+
+
+BATCH_SIZE = 1000
+TREE_QUERY = """\
+WITH RECURSIVE t(node, title, path, value, depth, corder, is_collated) AS (
+    SELECT nodeid, ARRAY[title], ARRAY[nodeid], documentid, 1, ARRAY[childorder],
+           is_collated
+    FROM trees tr, modules m
+    WHERE tr.documentid = m.module_ident
+          AND
+          tr.parent_id IS NULL
+          AND
+          tr.is_collated = TRUE
+UNION ALL
+    /* Recursion */
+    SELECT c1.nodeid,
+           /* concat the new record to the hierarchy of titles */
+           t.title || ARRAY[c1.title],
+           /* concat the new record to the path */
+           t.path || ARRAY[c1.nodeid],
+           c1.documentid,
+           t.depth+1,
+           t.corder || ARRAY[c1.childorder],
+           c1.is_collated
+    FROM trees c1 JOIN t ON (c1.parent_id = t.node)
+    WHERE NOT nodeid = ANY (t.path)
+          AND
+          t.is_collated = c1.is_collated
+)
+SELECT node, title
+FROM t LEFT JOIN modules m ON t.value = m.module_ident
+WINDOW w AS (ORDER BY corder)
+ORDER BY corder
+;
+"""
+
+# https://wiki.postgresql.org/wiki/Unnest_multidimensional_array
+CREATE_REDUCE_DIM = """\
+CREATE OR REPLACE FUNCTION public.reduce_dim(anyarray)
+RETURNS SETOF anyarray AS
+$function$
+DECLARE
+    s $1%TYPE;
+BEGIN
+    FOREACH s SLICE 1  IN ARRAY $1 LOOP
+        RETURN NEXT s;
+    END LOOP;
+    RETURN;
+END;
+$function$
+LANGUAGE plpgsql IMMUTABLE;
+"""
+
+DROP_REDUCE_DIM = """\
+DROP FUNCTION public.reduce_dim(anyarray);
+"""
+
+UPDATE_STMT = """\
+UPDATE trees SET slug = q.slug
+from (
+  SELECT yyy[1] AS id, yyy[2] AS slug FROM reduce_dim(%s) AS yyy
+) AS q
+WHERE nodeid = q.id::int
+;
+"""
+
+
+def batcher(seq, size):
+    for pos in range(0, len(seq), size):
+        yield seq[pos:pos + size]
+
+
+def generate_update_values(nodeid, title):
+    """Returns a sequence of trees.nodeid and trees.slug
+    to be used to update the trees slug table value.
+
+    """
+    logger.info("processing... {} - {}".format(nodeid, title))
+    try:
+        slug = generate_slug(*title)
+    except:
+        logger.exception(
+            "failed to create slug for '{}'".format(title)
+        )
+        raise
+    logger.info("... using {}".format(slug))
+    # must return an array of a single type for postgresql
+    return [str(nodeid), slug]
+
+
+def should_run(cursor):
+    # Regenerate the slugs as part of changing slug generation code per:
+    # https://github.com/openstax/cnx/issues/972
+    #
+    # We're specifically looking for HS Physics where slugs are have a prefix
+    # of "chapter" which we need to remove. We'll gate the run based on
+    # a known "chapter-2-motion-in-one-dimension" slug
+    cursor.execute("""\
+        SELECT slug FROM trees
+        JOIN modules ON trees.documentid = modules.module_ident
+        WHERE modules.uuid = 'd9956deb-accc-5d91-aa9a-81ad8053d29f'
```

Comment:
I randomly picked this module UUID since it's a slug we were looking at during deployment earlier. I'm not sure if this is the "right" way to hook onto things for `should_run`, but it did seem to work in practice when testing locally (e.g. the second migration detected it should run)